File: Leetcode/minimum_operations_to_make_a_uni-value_grid/min_operations.rb
# @param {Integer[][]} grid
# @param {Integer} x
# @return {Integer}
def min_operations(grid, x)
  flat = grid.flatten

  base = flat[0] % x
  return -1 unless flat.all? { |val| val % x == base }

  normalized = flat.map { |val| val / x }

  median = normalized.sort[normalized.size / 2]

  normalized.sum { |val| (val - median).abs }
end

p min_operations([[2,4],[6,8]], 2) #4
p min_operations([[1,5],[2,3]], 1) #5
p min_operations([[1,2],[3,4]], 2) #-1

File: Leetcode/minimum_operations_to_make_a_uni-value_grid/min_operations.go
package main

import (
	"fmt"
	"math"
	"sort"
)

func minOperations(grid [][]int, x int) int {
	// Flatten the grid
	flat := []int{}
	for _, row := range grid {
		for _, val := range row {
			flat = append(flat, val)
		}
	}

	// Check if it's possible to make all elements equal
	base := flat[0] % x
	for _, val := range flat {
		if val%x != base {
			return -1
		}
	}

	// Normalize the values
	for i := range flat {
		flat[i] = flat[i] / x
	}

	// Find the median
	sort.Ints(flat)
	median := flat[len(flat)/2]

	// Calculate total operations
	ops := 0
	for _, val := range flat {
		ops += int(math.Abs(float64(val - median)))
	}

	return ops
}

func main() {
	// Example inputs
	grid1 := [][]int{{2, 4}, {6, 8}}
	grid2 := [][]int{{1, 5}, {2, 3}}
	grid3 := [][]int{{1, 2}, {3, 4}}

	fmt.Println(minOperations(grid1, 2)) // Output: 4
	fmt.Println(minOperations(grid2, 1)) // Output: 5
	fmt.Println(minOperations(grid3, 2)) // Output: -1
}
